scanner/src/main/kotlin/experimental/ExperimentalScanner.kt

```diff
@@ -213,33 +214,69 @@ class ExperimentalScanner(
      * Run package scanners for packages with incomplete scan results.
      */
     private fun runPackageScanners(controller: ScanController, context: ScanContext) {
-        controller.packages.forEach { pkg ->
-            val nestedProvenance = controller.getNestedProvenance(pkg.id) ?: return@forEach
-
-            // TODO: Use coroutines to execute scanners in parallel.
+        controller.getPackagesConsolidatedByProvenance().forEach { (provenance, packages) ->
             controller.getPackageScanners().forEach scanner@{ scanner ->
-                if (controller.hasCompleteScanResult(scanner, pkg)) {
-                    log.debug {
-                        "Skipping scan of '${pkg.id.toCoordinates()}' with package scanner '${scanner.name}' as " +
-                                "stored results are available."
+                val packagesWithIncompleteScanResult = packages
```

Comment:
How about filtering only once like:

                val packagesWithIncompleteScanResult = packages.filter { pkg ->
                    val hasNestedProvenance = controller.getNestedProvenance(pkg.id) != null
                    if (!hasNestedProvenance) {
                        log.debug {
                            "Skipping scan of '${pkg.id.toCoordinates()}' with package scanner ${scanner.name} as no " +
                                    "nested provenance for the package could be resolved."
                        }
                    }

                    val hasCompleteScanResult = controller.hasCompleteScanResult(scanner, pkg)
                    if (hasCompleteScanResult) {
                        log.debug {
                            "Skipping scan of '${pkg.id.toCoordinates()}' with package scanner '${scanner.name}' as " +
                                    "stored results are available."
                        }
                    }

                    hasNestedProvenance && !hasCompleteScanResult
                }

Note that the above code also improves the log messages slightly.